Sources/SwiftSyntax/SyntaxChildren.swift

```diff
@@ -10,305 +10,246 @@
-struct RawSyntaxChildren: Sequence {
-  struct Iterator: IteratorProtocol {
-    let parent: RawSyntax
-    var nextChildInfo: AbsoluteSyntaxInfo
-
-    init(parent: AbsoluteRawSyntax) {
-      self.init(parent: parent.raw,
-        startFrom: parent.info.advancedToFirstChild())
-    }
-
-    init(parent: RawSyntax, startFrom: AbsoluteSyntaxInfo) {
-      self.parent = parent
-      self.nextChildInfo = startFrom
-    }
+public struct SyntaxChildrenIndex: Comparable {
+  fileprivate let offset: UInt32
+  fileprivate let indexInParent: UInt32
+  fileprivate let indexInTree: SyntaxIndexInTree
 
-    mutating func next() -> (RawSyntax?, AbsoluteSyntaxInfo)? {
-      let idx = Int(nextChildInfo.indexInParent)
-      guard idx < parent.numberOfChildren else {
-        return nil
-      }
-      let child = parent.child(at: idx)
-      let thisItem = (child, nextChildInfo)
-      nextChildInfo = nextChildInfo.advancedBySibling(child)
-      return thisItem
-    }
+  public static func <(lhs: SyntaxChildrenIndex, rhs: SyntaxChildrenIndex)
+      -> Bool {
+    return lhs.indexInParent < rhs.indexInParent
   }
 
-  private let node: AbsoluteRawSyntax
-
-  init(_ absoluteRaw: AbsoluteRawSyntax) {
-    self.node = absoluteRaw
+  fileprivate init(offset: UInt32, indexInParent: UInt32,
+                   indexInTree: SyntaxIndexInTree) {
+    self.offset = offset
+    self.indexInParent = indexInParent
+    self.indexInTree = indexInTree
   }
 
-  init(_ base: _SyntaxBase) {
-    self.init(base.data.absoluteRaw)
-  }
-
-  func makeIterator() -> Iterator {
-    return .init(parent: node)
+  init(_ absoluteSyntaxInfo: AbsoluteSyntaxInfo) {
+    self.offset = absoluteSyntaxInfo.offset
+    self.indexInParent = absoluteSyntaxInfo.indexInParent
+    self.indexInTree = absoluteSyntaxInfo.nodeId.indexInTree
   }
 }
 
-protocol AbsoluteRawSyntaxIteratorProtocol:
-  IteratorProtocol where Element == AbsoluteRawSyntax {
-}
+/// Collection that contains the child nodes of a raw node (including missing
+/// nodes), along with their associated `AbsoluteSyntaxInfo`.
+struct RawSyntaxChildren: BidirectionalCollection {
+  typealias Element = (node: RawSyntax?, info: AbsoluteSyntaxInfo)
+  typealias Index = SyntaxChildrenIndex
 
-/// Sequence of `AbsoluteRawSyntax` formed from all the `present` layout
-/// children nodes of a raw node.
-struct PresentRawSyntaxChildren: Sequence {
-  struct Iterator: AbsoluteRawSyntaxIteratorProtocol {
-    var iterator: RawSyntaxChildren.Iterator
+  /// The node whose children shall be accessed
+  private let parent: RawSyntax
+  private let rootId: UInt32
 
-    init(parent: AbsoluteRawSyntax) {
-      self.init(parent: parent.raw,
-        startFrom: parent.info.advancedToFirstChild())
-    }
-
-    init(parent: RawSyntax, startFrom: AbsoluteSyntaxInfo) {
-      self.iterator = .init(parent: parent, startFrom: startFrom)
-    }
+  let startIndex: SyntaxChildrenIndex
+  let endIndex: SyntaxChildrenIndex
```

Comment:
I think the `RawSyntaxChildren` struct size will be smaller if you just store the given `AbsoluteRawSyntax` and make the existing fields calculated properties. Calculating the `startIndex`/`endIndex` should be cheap.

Reply:
My idea was that recomputing the `endIndex` might have an ever so slight performance impact that occurs every time the collection is iterated which might happen multiple times.
Increasing the size of the collection by 32 bytes means we only need to allocate 32 more bytes on the stack once which should not make a noticeable performance difference since we are already allocating the space for the rest of the struct anyway.

Anyway, there should not be a noticable performance difference with either implementation, I believe.

Reply:
I'd suggest to measure both ways. For example, the visitation performance regression could be because `RawSyntaxChildren` is constructed all the time for every index iteration access for a `SyntaxCollection`, while if `RawSyntaxChildren` is just a wrapper of the existing `SyntaxCollection`'s `AbsoluteRawSyntax` then the optimizer may do a better job to remove overhead.